File: internal/cli/deinit.go
package cli

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/spf13/cobra"
)

var (
	deinitForce bool
)

var deinitCmd = &cobra.Command{
	Use:   "deinit",
	Short: "Remove Rafa from the current repository",
	Long:  "Removes the .rafa/ folder and all plans. This action cannot be undone.",
	RunE:  runDeinit,
}

func init() {
	deinitCmd.Flags().BoolVarP(&deinitForce, "force", "f", false, "Skip confirmation prompt")
}

func runDeinit(cmd *cobra.Command, args []string) error {
	// Check if initialized
	info, err := os.Stat(rafaDir)
	if os.IsNotExist(err) {
		return fmt.Errorf("rafa is not initialized in this repository")
	}
	if err != nil {
		return fmt.Errorf("failed to check .rafa directory: %w", err)
	}
	if !info.IsDir() {
		return fmt.Errorf(".rafa exists but is not a directory")
	}

	// Calculate what will be deleted
	planCount, totalSize, err := calculateDirStats(rafaDir)
	if err != nil {
		return fmt.Errorf("failed to analyze .rafa/: %w", err)
	}

	// Show confirmation unless --force
	if !deinitForce {
		fmt.Printf("This will delete .rafa/ (%d plans, %s). Continue? [y/N] ", planCount, formatSize(totalSize))

		reader := bufio.NewReader(os.Stdin)
		response, _ := reader.ReadString('\n')
		response = strings.TrimSpace(strings.ToLower(response))

		if response != "y" && response != "yes" {
			fmt.Println("Aborted.")
			return nil
		}
	}

	// Remove the directory
	if err := os.RemoveAll(rafaDir); err != nil {
		return fmt.Errorf("failed to remove .rafa/: %w", err)
	}

	// Remove lock file pattern from .gitignore
	if err := removeFromGitignore(gitignoreEntry); err != nil {
		return fmt.Errorf("failed to update .gitignore: %w", err)
	}

	fmt.Println("Rafa has been removed from this repository.")
	return nil
}

func calculateDirStats(dir string) (planCount int, totalSize int64, err error) {
	plansDir := filepath.Join(dir, "plans")
	entries, readErr := os.ReadDir(plansDir)
	if readErr == nil {
		planCount = len(entries)
	}

	err = filepath.Walk(dir, func(path string, info os.FileInfo, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if !info.IsDir() {
			totalSize += info.Size()
		}
		return nil
	})
	return
}

func formatSize(bytes int64) string {
	const (
		KB = 1024
		MB = KB * 1024
	)
	switch {
	case bytes >= MB:
		return fmt.Sprintf("%.1fMB", float64(bytes)/MB)
	case bytes >= KB:
		return fmt.Sprintf("%.1fKB", float64(bytes)/KB)
	default:
		return fmt.Sprintf("%dB", bytes)
	}
}

File: ts-port/src/cli/deinit.ts
/**
 * Deinit command - removes Rafa from the current repository
 *
 * Removes .rafa/ directory and uninstalls skills from .claude/skills/
 */

import * as fs from "node:fs/promises";
import * as path from "node:path";
import * as readline from "node:readline/promises";
import { stdin, stdout } from "node:process";
import { uninstallSkills } from "../core/skills.js";
import { isInitialized, getRafaDir, getSkillsDir, removeFromGitignore } from "./init.js";

const GITIGNORE_LOCK_ENTRY = ".rafa/**/*.lock";
const GITIGNORE_SESSIONS_ENTRY = ".rafa/sessions/";

/**
 * Options for the deinit command
 */
export interface DeinitOptions {
  workDir?: string;
  force?: boolean; // Skip confirmation prompt
  confirm?: () => Promise<boolean>; // Custom confirmation function (for testing)
}

/**
 * Result of the deinit command
 */
export interface DeinitResult {
  success: boolean;
  message: string;
}

/**
 * Calculates directory statistics
 */
async function calculateDirStats(
  dir: string
): Promise<{ planCount: number; totalSize: number }> {
  let planCount = 0;
  let totalSize = 0;

  // Count plans
  const plansDir = path.join(dir, "plans");
  try {
    const entries = await fs.readdir(plansDir);
    planCount = entries.length;
  } catch {
    // Ignore if plans directory doesn't exist
  }

  // Calculate total size
  async function walkDir(currentDir: string): Promise<void> {
    try {
      const entries = await fs.readdir(currentDir, { withFileTypes: true });
      for (const entry of entries) {
        const fullPath = path.join(currentDir, entry.name);
        if (entry.isDirectory()) {
          await walkDir(fullPath);
        } else {
          try {
            const stat = await fs.stat(fullPath);
            totalSize += stat.size;
          } catch {
            // Ignore files we can't stat
          }
        }
      }
    } catch {
      // Ignore directories we can't read
    }
  }

  await walkDir(dir);

  return { planCount, totalSize };
}

/**
 * Formats a size in bytes to a human-readable string
 */
function formatSize(bytes: number): string {
  const KB = 1024;
  const MB = KB * 1024;

  if (bytes >= MB) {
    return `${(bytes / MB).toFixed(1)}MB`;
  } else if (bytes >= KB) {
    return `${(bytes / KB).toFixed(1)}KB`;
  } else {
    return `${bytes}B`;
  }
}

/**
 * Prompts the user for confirmation
 */
async function promptConfirmation(message: string): Promise<boolean> {
  const rl = readline.createInterface({ input: stdin, output: stdout });
  try {
    const answer = await rl.question(message);
    const response = answer.trim().toLowerCase();
    return response === "y" || response === "yes";
  } finally {
    rl.close();
  }
}

/**
 * Runs the deinit command
 */
export async function runDeinit(options: DeinitOptions = {}): Promise<DeinitResult> {
  const workDir = options.workDir ?? process.cwd();
  const rafaDir = getRafaDir(workDir);
  const skillsDir = getSkillsDir(workDir);

  // Check if initialized
  if (!(await isInitialized(workDir))) {
    return {
      success: false,
      message: "Rafa is not initialized in this repository",
    };
  }

  // Verify .rafa is a directory
  try {
    const stat = await fs.stat(rafaDir);
    if (!stat.isDirectory()) {
      return {
        success: false,
        message: ".rafa exists but is not a directory",
      };
    }
  } catch (err) {
    return {
      success: false,
      message: `Failed to check .rafa directory: ${err instanceof Error ? err.message : err}`,
    };
  }

  // Calculate what will be deleted
  const { planCount, totalSize } = await calculateDirStats(rafaDir);

  // Show confirmation unless --force
  if (!options.force) {
    const confirmFn = options.confirm ?? promptConfirmation;
    const confirmed = await confirmFn(
      `This will delete .rafa/ (${planCount} plans, ${formatSize(totalSize)}) and remove skills from .claude/skills/. Continue? [y/N] `
    );

    if (!confirmed) {
      return {
        success: true,
        message: "Aborted.",
      };
    }
  }

  // Remove skills
  try {
    await uninstallSkills(skillsDir);
  } catch (err) {
    console.warn(
      `Warning: failed to remove skills: ${err instanceof Error ? err.message : err}`
    );
  }

  // Remove the directory
  try {
    await fs.rm(rafaDir, { recursive: true, force: true });
  } catch (err) {
    return {
      success: false,
      message: `Failed to remove .rafa/: ${err instanceof Error ? err.message : err}`,
    };
  }

  // Remove gitignore entries
  try {
    await removeFromGitignore(GITIGNORE_LOCK_ENTRY, workDir);
    await removeFromGitignore(GITIGNORE_SESSIONS_ENTRY, workDir);
  } catch (err) {
    return {
      success: false,
      message: `Failed to update .gitignore: ${err instanceof Error ? err.message : err}`,
    };
  }

  return {
    success: true,
    message:
      "Rafa has been removed from this repository (skills uninstalled from .claude/skills/).",
  };
}
